osbuild/monitor.py

```diff
@@ -9,17 +9,175 @@
 """
 
 import abc
+import copy
 import datetime
+import hashlib
 import json
 import os
 import sys
 import time
-from typing import Dict
+from typing import Dict, Optional, Set
 
 import osbuild
 from osbuild.util.term import fmt as vt
 
 
+def omitempty(d: dict):
+    """ Omit None and empty string ("") values from the given dict """
+    for k, v in list(d.items()):
+        if v is None or v == "":
+            del d[k]
+        elif isinstance(v, dict):
+            omitempty(v)
+    return d
+
+
+class Context:
+    """Context for a single log entry. Automatically calculates hash/id when read."""
+
+    def __init__(self,
+                 origin: Optional[str] = None,
+                 pipeline: Optional[osbuild.Pipeline] = None,
+                 stage: Optional[osbuild.Stage] = None):
+        self._origin = origin
+        self._pipeline_name = pipeline.name if pipeline else None
+        self._pipeline_id = pipeline.id if pipeline else None
+        self._stage_name = stage.name if stage else None
+        self._stage_id = stage.id if stage else None
+        self._id = None
+        self._id_history: Set[str] = set()
+
+    def __setattr__(self, name, value):
+        super().__setattr__(name, value)
+        # reset "_id" on any write so that the hash is automatically recalculated
+        if name != "_id":
+            super().__setattr__("_id", None)
+
+    def with_origin(self, origin: Optional[str]) -> "Context":
+        """
+        Return a Context with the given origin but otherwise identical.
+
+        Note that if the origin is empty or same it will return self.
+        """
+        if origin is None or origin == self._origin:
+            return self
+        ctx = copy.copy(self)
+        ctx.set_origin(origin)
+        return ctx
+
+    @property
```

Comment:
For these properties (and the following) we could also use `@property.setter` instead of having explicit `set_x` methods.

Reply:
I like this, I updated "origin". For the others it's slightly tricky because there is no `self._pipeline` so no `@property\npipeline` only pipline_{name,id}. This is from the original code from @achilleas-k - I assume to keep the memory overhead for the context small(?) so I'm hesitant to change it). I hope I'm making sense.